File: iOS/KajHobe/Managers/MessageBadgeManager.swift
import SwiftUI
import Combine
import Supabase

/// Manages the unread message badge count on the Messages tab.
/// Tracks the total count of unread messages across ALL conversations
/// for the current user, updated in real-time via Supabase subscriptions.
class MessageBadgeManager: ObservableObject {
    static let shared = MessageBadgeManager()

    @Published var totalUnreadCount: Int = 0

    private var realtimeChannel: RealtimeChannelV2?
    private var currentUserId: String?

    private init() {
        setupRealtimeSubscription()
    }

    deinit {
        Task { [weak realtimeChannel] in
            await realtimeChannel?.unsubscribe()
        }
    }

    // MARK: - Public Methods

    /// Initial fetch of unread count. Called on login/launch.
    func refreshCounts() async {
        do {
            let user = try await supabase.auth.user()
            currentUserId = user.id.uuidString
            await fetchUnreadCount(userId: user.id.uuidString)
        } catch {
            print("❌ Error refreshing message counts: \(error)")
        }
    }

    /// Decrement the badge by the number of messages just marked as read in a conversation.
    /// Called from ChatView after marking messages as read.
    func decrement(by count: Int) {
        guard count > 0 else { return }
        DispatchQueue.main.async {
            self.totalUnreadCount = max(0, self.totalUnreadCount - count)
        }
    }

    /// Reset count to a known value (e.g., after conversation opened). Clamps to 0.
    func setCount(_ count: Int) {
        DispatchQueue.main.async {
            self.totalUnreadCount = max(0, count)
        }
    }

    // MARK: - Private Methods

    private func fetchUnreadCount(userId: String) async {
        do {
            let response = try await supabase
                .from("messages")
                .select("id", count: .exact)
                .is("read_at", value: nil)
                .neq("sender_id", value: userId)
                .execute()
            let count = response.count ?? 0
            await MainActor.run {
                self.totalUnreadCount = count
            }
            print("💬 Message badge: \(count) unread")
        } catch {
            print("⚠️ Could not fetch unread message count: \(error)")
        }
    }

    private func setupRealtimeSubscription() {
        Task {
            do {
                let user = try await supabase.auth.user()
                let userId = user.id.uuidString
                currentUserId = userId

                // Subscribe to public:messages — same channel the rest of the messaging
                // code uses, so a single subscription serves both list and badge updates.
                let channel = supabase.realtimeV2.channel("public:messages")

                // Listen for INSERTs: new message from another user → +1 to badge.
                let insertions = await channel.postgresChange(InsertAction.self, table: "messages")

                // Listen for UPDATEs: read_at changed from NULL to a value → -1 to badge
                // (only when the message wasn't from us and was previously unread).
                let updates = await channel.postgresChange(UpdateAction.self, table: "messages")

                realtimeChannel = channel
                await channel.subscribe()

                // Initial count.
                await fetchUnreadCount(userId: userId)

                // Spawn collectors.
                Task { [weak self] in
                    for await insertion in insertions {
                        await self?.handleInsertion(insertion)
                    }
                }
                Task { [weak self] in
                    for await update in updates {
                        await self?.handleUpdate(update)
                    }
                }

                print("💬 MessageBadgeManager subscribed for user: \(userId)")
            } catch {
                print("❌ Error setting up MessageBadgeManager subscription: \(error)")
            }
        }
    }

    private func handleInsertion(_ action: HasRecord) async {
        let record = action.record
        guard let senderId = record["sender_id"]?.stringValue,
              let readAt = record["read_at"] else {
            return
        }
        guard let uid = currentUserId else { return }
        // Only count messages from OTHER users; only count if currently unread.
        guard senderId != uid else { return }
        if let readAtStr = readAt.stringValue, !readAtStr.isEmpty {
            return // already read, no badge increment
        }
        await MainActor.run {
            self.totalUnreadCount += 1
            print("💬 Message badge +1 → \(self.totalUnreadCount)")
        }
    }

    private func handleUpdate(_ action: HasRecord) async {
        let record = action.record
        guard let readAt = record["read_at"]?.stringValue,
              let senderId = record["sender_id"]?.stringValue else {
            return
        }
        guard let uid = currentUserId else { return }
        // Only decrement when one of OUR messages was just marked read by the recipient.
        guard senderId == uid, !readAt.isEmpty else { return }
        await MainActor.run {
            self.totalUnreadCount = max(0, self.totalUnreadCount - 1)
            print("💬 Message badge -1 → \(self.totalUnreadCount)")
        }
    }
}

// MARK: - SwiftUI Environment Key
struct MessageBadgeManagerKey: EnvironmentKey {
    static let defaultValue = MessageBadgeManager.shared
}

extension EnvironmentValues {
    var messageBadgeManager: MessageBadgeManager {
        get { self[MessageBadgeManagerKey.self] }
        set { self[MessageBadgeManagerKey.self] = newValue }
    }
}

File: Android/app/src/main/java/com/kajhobe/app/data/notifications/MessageBadgeManager.kt
package com.kajhobe.app.data.notifications

import com.kajhobe.app.data.model.ChatMessage
import io.github.jan.supabase.SupabaseClient
import io.github.jan.supabase.auth.auth
import io.github.jan.supabase.postgrest.postgrest
import io.github.jan.supabase.postgrest.query.Columns
import io.github.jan.supabase.realtime.PostgresAction
import io.github.jan.supabase.realtime.RealtimeChannel
import io.github.jan.supabase.realtime.channel
import io.github.jan.supabase.realtime.decodeRecord
import io.github.jan.supabase.realtime.postgresChangeFlow
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.serialization.Serializable

/**
 * Messages tab badge — total unread message count across all conversations for
 * the current user, kept fresh in real-time via the same `public:messages` channel
 * the conversation list uses. Android port of iOS `MessageBadgeManager`.
 */
class MessageBadgeManager(
    private val client: SupabaseClient,
) {
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val postgrest get() = client.postgrest
    private val auth get() = client.auth

    private val _totalUnreadCount = MutableStateFlow(0)
    val totalUnreadCount: StateFlow<Int> = _totalUnreadCount.asStateFlow()

    private var channel: RealtimeChannel? = null
    private var insertJob: Job? = null
    private var updateJob: Job? = null
    private var currentUserId: String? = null

    init {
        // Boot the realtime subscription as soon as the manager is created.
        // start() re-checks the user each time so the subscription is rebound on login.
        scope.launch { start() }
    }

    /**
     * (Re)bind the subscription to the current user. Safe to call on login.
     *
     * IMPORTANT ordering rule (supabase-kt): `postgresChangeFlow` MUST be called
     * before the channel is joined (`channel.subscribe()`). Calling it after a
     * channel with the same name has already been subscribed elsewhere will
     * throw `IllegalStateException: You cannot call postgresChangeFlow after
     * joining the channel`. Because `client.channel("public:messages")` is
     * de-duped by RealtimeImpl, the conversations view and this manager SHARE
     * the same channel — so we must register listeners BEFORE anyone else
     * subscribes.
     *
     * Strategy: this manager runs at app startup (Koin singleton), well before
     * the user opens the Messages tab. We register our listeners here and DO
     * NOT call `subscribe()` — that is the conversations view's job. As long
     * as our `postgresChangeFlow` calls run before the conversations view's
     * listener registration, the shared channel will join with all listeners
     * attached and the messages will flow to us.
     */
    suspend fun start() {
        val uid = auth.currentUserOrNull()?.id
        if (uid == null) {
            _totalUnreadCount.value = 0
            return
        }
        if (uid == currentUserId && channel != null) return // already bound
        currentUserId = uid

        // Tear down any prior subscription before rebuilding.
        stopInternal()

        // Initial count (one SELECT) — fire and forget.
        scope.launch { refreshCount() }

        val ch = client.channel("public:messages")
        channel = ch

        // Register BOTH listeners synchronously. `postgresChangeFlow` calls
        // `addPostgresChange` immediately at call time, so as long as we invoke
        // these BEFORE the channel is in SUBSCRIBED state, we're safe. The
        // conversations view will eventually call subscribe() — at that point
        // we'll already be wired up.
        val insertFlow = ch.postgresChangeFlow<PostgresAction.Insert>(schema = "public") {
            table = "messages"
        }
        val updateFlow = ch.postgresChangeFlow<PostgresAction.Update>(schema = "public") {
            table = "messages"
        }

        // Collect the two flows. Listeners are already attached at this point.
        insertJob = scope.launch {
            insertFlow.collect { action ->
                val msg = runCatching { action.decodeRecord<ChatMessage>() }.getOrNull() ?: return@collect
                handleInsert(msg)
            }
        }
        updateJob = scope.launch {
            updateFlow.collect { action ->
                val msg = runCatching { action.decodeRecord<ChatMessage>() }.getOrNull() ?: return@collect
                handleUpdate(msg)
            }
        }

        // NOTE: We do NOT call ch.subscribe() here. The ConversationsViewModel
        // owns the channel lifecycle for "public:messages" — calling subscribe
        // twice is a no-op, but it would also mean our listeners get registered
        // only AFTER someone else subscribes (a race). Leaving it to the
        // conversations view guarantees the order: register-listener → subscribe.
    }

    /** Stop the realtime subscription (e.g., on sign-out). */
    suspend fun stop() {
        currentUserId = null
        stopInternal()
        _totalUnreadCount.value = 0
    }

    /** Optimistic local decrement (used by the chat screen after marking messages read). */
    fun decrement(by: Int) {
        if (by <= 0) return
        _totalUnreadCount.update { (it - by).coerceAtLeast(0) }
    }

    /** Force a full re-fetch from the server. */
    fun refreshCounts() {
        scope.launch { refreshCount() }
    }

    private suspend fun refreshCount() {
        val uid = currentUserId ?: return
        runCatching {
            // Just selecting `id` is enough — we only care about the list size.
            postgrest.from("messages")
                .select(Columns.list("id")) {
                    filter { exact("read_at", null); neq("sender_id", uid) }
                }
                .decodeList<UnreadIdRow>()
                .size
        }.onSuccess { count ->
            _totalUnreadCount.value = count
        }
    }

    private fun handleInsert(msg: ChatMessage) {
        val uid = currentUserId ?: return
        // Only count messages from OTHER users; skip if it was already read.
        if (msg.sender_id == uid) return
        if (!msg.read_at.isNullOrEmpty()) return
        _totalUnreadCount.update { it + 1 }
    }

    private fun handleUpdate(msg: ChatMessage) {
        val uid = currentUserId ?: return
        // A previously-unread message from the other party just got a read_at → -1.
        // (We don't have the prior state here, so we only decrement when read_at is now set
        // and the sender is the other party — this matches the "marked as read" signal.)
        if (msg.sender_id == uid) return // we sent it; the recipient reading it has no badge effect
        if (msg.read_at.isNullOrEmpty()) return
        _totalUnreadCount.update { (it - 1).coerceAtLeast(0) }
    }

    private suspend fun stopInternal() {
        insertJob?.cancel(); insertJob = null
        updateJob?.cancel(); updateJob = null
        // We never subscribe the channel (the conversations view owns the
        // subscription lifecycle for "public:messages"), so there's nothing to
        // unsubscribe. We just drop our reference to the channel so the next
        // start() can re-register its listeners cleanly.
        channel = null
    }

    @Serializable
    private data class UnreadIdRow(val id: String? = null)
}
